### PR TITLE
Use broadcast_like for 2d plot coordinates

### LINKED ISSUE #5097: 2d plots may fail for some choices of `x` and `y`
**What happened**:
When making a 2d plot with a 1d `x` argument and a 2d `y`, if the two dimensions have the same size and are in the wrong order, no plot is produced - the third plot in the MCVE is blank.

**What you expected to happen**:
All three plots in the MCVE should be identical.

**Minimal Complete Verifiable Example**:

```python
from matplotlib import pyplot as plt
import numpy as np
import xarray as xr

ds = xr.Dataset({"z": (["x", "y"], np.random.rand(4,4))})

x2d, y2d = np.meshgrid(ds["x"], ds["y"])

ds = ds.assign_coords(x2d=(["x", "y"], x2d.T), y2d=(["x", "y"], y2d.T))

fig, axes = plt.subplots(1,3)

h0 = ds["z"].plot.pcolormesh(x="y2d", y="x2d", ax=axes[0])
h1 = ds["z"].plot.pcolormesh(x="y", y="x", ax=axes[1])
h2 = ds["z"].plot.pcolormesh(x="y", y="x2d", ax=axes[2])

plt.show()
```

result:
test2d

**Anything else we need to know?**:

The bug is present in both the 0.17.0 release and current `master`.

I came across this while starting to work on #5084. I think the problem is here
https://github.com/pydata/xarray/blob/ddc352faa6de91f266a1749773d08ae8d6f09683/xarray/plot/plot.py#L678-L684
as the check `xval.shape[0] == yval.shape[0]` doesn't work if the single dimension of x is actually the second dimension of y, but happened to have the same size as the first dimension of y? I think it needs to check the actual dimensions of `x` and `y`.

Why don't we just do something like
```
xval = xval.broadcast_like(darray)
yval = yval.broadcast_like(darray)
```
if either coordinate is 2d before using `.values` to convert to numpy arrays?

**Environment**:

<details><summary>Output of <tt>xr.show_versions()</tt></summary>

INSTALLED VERSIONS
------------------
commit: None
python: 3.8.6 | packaged by conda-forge | (default, Oct  7 2020, 19:08:05) 
[GCC 7.5.0]
python-bits: 64
OS: Linux
OS-release: 5.4.0-70-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_GB.UTF-8
LOCALE: en_GB.UTF-8
libhdf5: 1.10.6
libnetcdf: 4.7.4

xarray: 0.17.0
pandas: 1.1.5
numpy: 1.19.4
scipy: 1.5.3
netCDF4: 1.5.5.1
pydap: None
h5netcdf: None
h5py: 3.1.0
Nio: None
zarr: None
cftime: 1.3.0
nc_time_axis: None
PseudoNetCDF: None
rasterio: None
cfgrib: None
iris: None
bottleneck: None
dask: 2020.12.0
distributed: 2020.12.0
matplotlib: 3.3.3
cartopy: None
seaborn: None
numbagg: None
pint: 0.16.1
setuptools: 49.6.0.post20201009
pip: 20.3.3
conda: 4.9.2
pytest: 6.2.1
IPython: 7.19.0
sphinx: 3.4.0

</details>

### PR DESCRIPTION
**Pull request by johnomotani**
Use broadcast_like if either `x` or `y` inputs are 2d to ensure that both have dimensions in the same order as the DataArray being plotted. Convert to numpy arrays after possibly using broadcast_like. Simplifies code, and fixes #5097 (bug when dimensions have the same size).

@dcherian
> IIRC the "resolving intervals" section later will break and is somewhat annoying to fix. This is why the current ugly code exists.

This change seems to 'just work', and unit tests pass. Is there some extra check that needs doing to make sure "resolving intervals" is behaving correctly?

I can't think of a unit test that would have caught #5097, since even when the bug happens, a plot is produced without errors or warnings. If anyone has an idea, suggestions/pushes welcome!

- [x] Closes #5097
- [ ] Tests added
- [x] Passes `pre-commit run --all-files`
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

### CODE REVIEW

**Comment by dcherian** [comment]
> This change seems to 'just work', and unit tests pass. Is there some extra check that needs doing to make sure "resolving intervals" is behaving correctly?

OK I must be misremembering then.

**Review by mathause — COMMENTED** [review]
Thanks for your patience! In your MVCE (#5097) compare `h1.get_paths()[0]` and h`2.get_paths()[0]`. Thus, you can use `h2.get_paths()` for a test e.g.:

```python
v = h2.get_paths()[0].vertices
assert not np.all(v[:, 0] == v[:, 1])
```
(in the MVCE all x and y coords are equal i.e. it's a diagonal with width 0).

I suggest merging this (pending a test if you are up to it), and if something breaks the interval resolution that is not tested we fix it later...

**Comment by johnomotani** [comment]
Thanks @mathause! I've used your suggestion to turn my MVCE into a test which fails on master, but passes with the fix in this PR.

**Review by mathause — APPROVED** [review]
LGTM. Let's see if someone else wants to comment and I'll merge in a day or two.

File: doc/whats-new.rst [hunk 1]
```diff
@@ -78,6 +78,9 @@ Deprecations
 
 Bug fixes
 ~~~~~~~~~
+- Fix 2d plot failure for certain combinations of dimensions when `x` is 1d and `y` is
+  2d (:issue:`5097`, :pull:`5099`).
+  By `John Omotani <https://github.com/johnomotani>`_
```
Comment (mathause) [inline doc/whats-new.rst:83]:
```suggestion
  2d (:issue:`5097`, :pull:`5099`). By `John Omotani <https://github.com/johnomotani>`_.
```

**Comment by mathause** [comment]
Thanks @johnomotani - fingers crossed that this does not break anything.

**Pull request merged**